Bottom: dc38ee19ceaac265b8c4bf884db9e9a694a401c2
Top:    0af6320e3d3fda58b6af0a8381874c348d3675a0
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-12-04 20:31:11 +0000

Fixup SearchObjectByLocation not to use negative layer numbers


---

diff --git a/src/search.c b/src/search.c
index 6c00673..bf7413a 100644
--- a/src/search.c
+++ b/src/search.c
@@ -1129,6 +1129,74 @@ IsPointOnArc (Coord X, Coord Y, Coord Radius, ArcType *Arc)
   return fabs (Distance (X, Y, Arc->X, Arc->Y) - Arc->Width) < Radius + Arc->Thickness / 2;
 }
 
+static unsigned int
+SearchLayerObjectInternal (LayerType *layer,
+                           unsigned int Type,
+                           unsigned int HigherAvail,
+                           double HigherBound,
+                           void **Result1,
+                           void **Result2,
+                           void **Result3)
+{
+  int locked = Type & LOCKED_TYPE;
+
+  if (!layer->On)
+    return NO_TYPE;
+
+  SearchLayer = layer;
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
+      Type & POLYGONPOINT_TYPE &&
+      SearchPointByLocation (locked, (LayerType **)Result1, (PolygonType **) Result2, (PointType **) Result3))
+    return POLYGONPOINT_TYPE;
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
+      Type & LINEPOINT_TYPE &&
+      SearchLinePointByLocation (locked, (LayerType **)Result1, (LineType **) Result2, (PointType **) Result3))
+    return LINEPOINT_TYPE;
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & LINE_TYPE
+      && SearchLineByLocation (locked, (LayerType **)Result1, (LineType **) Result2, (LineType **) Result3))
+    return LINE_TYPE;
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
+      Type & ARCPOINT_TYPE &&
+      SearchArcPointByLocation (locked, (LayerType **)Result1, (ArcType **) Result2, (PointType **) Result3))
+    return ARCPOINT_TYPE;
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & ARC_TYPE &&
+      SearchArcByLocation (locked, (LayerType **)Result1, (ArcType **) Result2, (ArcType **) Result3))
+    return ARC_TYPE;
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & TEXT_TYPE
+      && SearchTextByLocation (locked, (LayerType **)Result1, (TextType **) Result2, (TextType **) Result3))
+    return TEXT_TYPE;
+
+  if (Type & POLYGON_TYPE &&
+      SearchPolygonByLocation (locked, (LayerType **)Result1, (PolygonType **) Result2, (PolygonType **) Result3))
+    {
+      if (HigherAvail)
+        {
+          BoxType *box =
+            &(*(PolygonType **) Result2)->BoundingBox;
+          double area =
+            (double) (box->X2 - box->X1) * (double) (box->X2 - box->X1);
+          /* XXX: BEHAVIOURAL CHANGE */
+          if (HigherBound < area)
+//            break;
+            return NO_TYPE;
+          else
+            return POLYGON_TYPE;
+        }
+      else
+        {
+          return POLYGON_TYPE;
+        }
+    }
+
+  return NO_TYPE;
+}
+
 /* ---------------------------------------------------------------------------
  * searches for any kind of object or for a set of object types
  * the calling routine passes two pointers to allocated memory for storing
@@ -1152,8 +1220,9 @@ SearchObjectByLocation (unsigned Type,
   void **pr1 = &r1, **pr2 = &r2, **pr3 = &r3;
   int i;
   double HigherBound = 0;
-  int HigherAvail = NO_TYPE;
+  unsigned int HigherAvail = NO_TYPE;
   int locked = Type & LOCKED_TYPE;
+  unsigned int type;
   /* setup variables used by local functions */
   PosX = X;
   PosY = Y;
@@ -1234,95 +1303,32 @@ SearchObjectByLocation (unsigned Type,
       HigherAvail = ELEMENT_TYPE;
     }
 
-  for (i = -2; i < max_copper_layer + 2; i++)
-    {
-      if (i == -2)
-	SearchLayer = &PCB->Data->SOLDERMASKLAYER;
-      else if (i == -1)
-	SearchLayer = &PCB->Data->SILKLAYER;
-      else if (i < max_copper_layer)
-	SearchLayer = LAYER_ON_STACK (i);
-      else if (i == max_copper_layer)
-	{
-	  SearchLayer = &PCB->Data->BACKSILKLAYER;
-	  if (!PCB->InvisibleObjectsOn)
-	    continue;
-	}
-      else if (i == max_copper_layer + 1)
-	{
-	  SearchLayer = &PCB->Data->BACKSOLDERMASKLAYER;
-	  if (!PCB->InvisibleObjectsOn)
-	    continue;
-	}
-      if (SearchLayer->On)
-	{
-	  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
-	      Type & POLYGONPOINT_TYPE &&
-	      SearchPointByLocation (locked,
-				     (LayerType **) Result1,
-				     (PolygonType **) Result2,
-				     (PointType **) Result3))
-	    return (POLYGONPOINT_TYPE);
-
-	  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
-	      Type & LINEPOINT_TYPE &&
-	      SearchLinePointByLocation (locked,
-					 (LayerType **) Result1,
-					 (LineType **) Result2,
-					 (PointType **) Result3))
-	    return (LINEPOINT_TYPE);
+  type = SearchLayerObjectInternal (&PCB->Data->SOLDERMASKLAYER, Type, HigherAvail, HigherBound, Result1, Result2, Result3);
+  if (type != NO_TYPE)
+    return type;
 
-	  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & LINE_TYPE
-	      && SearchLineByLocation (locked,
-				       (LayerType **) Result1,
-				       (LineType **) Result2,
-				       (LineType **) Result3))
-	    return (LINE_TYPE);
+  type = SearchLayerObjectInternal (&PCB->Data->SILKLAYER, Type, HigherAvail, HigherBound, Result1, Result2, Result3);
+  if (type != NO_TYPE)
+    return type;
 
-	    if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
-	      Type & ARCPOINT_TYPE &&
-	      SearchArcPointByLocation (locked,
-					(LayerType **) Result1,
-					(ArcType **) Result2,
-					(PointType **) Result3))
-	    return (ARCPOINT_TYPE);
-
-	  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & ARC_TYPE &&
-	      SearchArcByLocation (locked,
-				   (LayerType **) Result1,
-				   (ArcType **) Result2,
-				   (ArcType **) Result3))
-	    return (ARC_TYPE);
+  for (i = 0; i < max_copper_layer; i++)
+    {
+      type = SearchLayerObjectInternal (LAYER_ON_STACK (i), Type, HigherAvail, HigherBound, Result1, Result2, Result3);
+      if (type != NO_TYPE)
+        return type;
+    }
 
-	  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & TEXT_TYPE
-	      && SearchTextByLocation (locked,
-				       (LayerType **) Result1,
-				       (TextType **) Result2,
-				       (TextType **) Result3))
-	    return (TEXT_TYPE);
+  if (PCB->InvisibleObjectsOn)
+    {
+      type = SearchLayerObjectInternal (&PCB->Data->BACKSILKLAYER, Type, HigherAvail, HigherBound, Result1, Result2, Result3);
+      if (type != NO_TYPE)
+        return type;
 
-	  if (Type & POLYGON_TYPE &&
-	      SearchPolygonByLocation (locked,
-				       (LayerType **) Result1,
-				       (PolygonType **) Result2,
-				       (PolygonType **) Result3))
-	    {
-	      if (HigherAvail)
-		{
-		  BoxType *box =
-		    &(*(PolygonType **) Result2)->BoundingBox;
-		  double area =
-		    (double) (box->X2 - box->X1) * (double) (box->X2 - box->X1);
-		  if (HigherBound < area)
-		    break;
-		  else
-		    return (POLYGON_TYPE);
-		}
-	      else
-		return (POLYGON_TYPE);
-	    }
-	}
+      type = SearchLayerObjectInternal (&PCB->Data->BACKSOLDERMASKLAYER, Type, HigherAvail, HigherBound, Result1, Result2, Result3);
+      if (type != NO_TYPE)
+        return type;
     }
+
   /* return any previously found objects */
   if (HigherAvail & PIN_TYPE)
     {
